tap-version-14-specification.md

```diff
@@ -0,0 +1,741 @@
+---
+layout: default
+title: TAP 14 specification
+---
+
+# NAME
+
+TAP14 - The Test Anything Protocol v14
+
+## SYNOPSIS
+
+TAP, the Test Anything Protocol, is a simple text-based interface
+between testing modules a test harness. TAP started life as part of
+the test harness for Perl but now has implementations in C/C++,
+Python, PHP, Perl, JavaScript, MatLab, and probably others by the time
+you read this.
+
+This document describes version 14 of TAP.
+
+## THE TAP14 FORMAT
+
+A TAP stream is a UTF-8 encoded text stream representing a set of
+tests.  It should be parsed line by line.  Lines are separated by
+`\n`.
+
+TAP14's general format is:
+
+```
+TAPDocument ->
+  Version Plan Body |
+  Version Body Plan
+```
+
+The Version is always `TAP version 14`.
+
+The Body is a collection of lines representing a test set.
+
+The Plan reports on the number of tests included in the Body.
+
+Any line matching the pattern `/^\s*#.*$/` is treated as a diagnostic
+comment, and MAY be reported to the user or ignored, but MUST be
+ignored for the purposes of interpreting test output semantics.
+
+For example:
+
+```
+TAP version 14
+1..4
+ok 1 Description # Directive
+# Diagnostic
+  ---
+  message: 'Failure message'
+  severity: fail
+  data:
+    found:
+      - 1
+      - 3
+      - 2
+    wanted:
+      - 1
+      - 2
+      - 3
+  ...
+ok 2 Description
+ok 3 Description
+    # Subtest: name of the child test
+    ok 1 this is ok
+    ok 2 this is also ok
+    not ok 3 this is a failure
+      ---
+      message: 'Child Failure'
+      severity: fail
+      ...
+    1..3
+not ok 4 name of the child test
+```
+
+For example, a test file's output might look like:
+
+```
+TAP version 14
+1..4
+ok 1 - Input file opened
+not ok 2 - First line of the input valid
+  ---
+  message: 'First line invalid'
+  severity: fail
+  data:
+    got: 'Flirble'
+    expect: 'Fnible'
+  ...
+ok 3 - Read the rest of the file
+not ok 4 - Summarized correctly # TODO Not written yet
+  ---
+  message: "Can't make summary yet"
+  severity: todo
+  ...
+```
+
+## HARNESS BEHAVIOR
+
+In this document, the "harness" is any program analyzing TAP output.
+
+A harness interpreting a subprocess MUST only read TAP output from
+standard output and not from standard error.
+
+Lines written to standard output matching `/^(not)? ok\b/` must be
+interpreted as test lines. After a test line a block of lines starting
+with `/^ {2}---$/` and ending with `/^ {2}\.\.\.$/` will be interpreted as an
+inline YAML document providing extended diagnostic information about
+the preceding test.
+
+All other lines must not be considered test output.
+
+## TESTS LINES AND THE PLAN
+
+### The Version
+
+To indicate that this is TAP14 the first non-comment, non-whitespace
```

Comment:
Can this include "non-unknown" too? Considering producers that log other random stuff that is hard to control, it might help a producer to conform easier without being ambiguous. For instance, `ember test` output can look like:

```
version: 1.13.8
Could not find watchman, falling back to NodeWatcher for file system events.
Visit http://www.ember-cli.com/user-guide/#watchman for more info.
Built project successfully. Stored in "/home/matt/testery/client/tmp/class-tests_dist-OyaYxzuN.tmp".
ok 1 PhantomJS 1.9 - JSHint - .: app.js should pass jshint
ok 2 PhantomJS 1.9 - JSHint - helpers: helpers/resolver.js should pass jshint
ok 3 PhantomJS 1.9 - JSHint - helpers: helpers/start-app.js should pass jshint
ok 4 PhantomJS 1.9 - JSHint - models: models/build.js should pass jshint
ok 5 PhantomJS 1.9 - JSHint - .: router.js should pass jshint
ok 6 PhantomJS 1.9 - JSHint - routes: routes/builds.js should pass jshint
ok 7 PhantomJS 1.9 - JSHint - .: test-helper.js should pass jshint
ok 8 PhantomJS 1.9 - Unit | Model | build: it exists
ok 9 PhantomJS 1.9 - Unit | Model | build: it has passes
ok 10 PhantomJS 1.9 - JSHint - unit/models: unit/models/build-test.js should pass jshint
ok 11 PhantomJS 1.9 - Unit | Route | builds: it exists
ok 12 PhantomJS 1.9 - JSHint - unit/routes: unit/routes/builds-test.js should pass jshint

1..12
# tests 12
# pass  12
# fail  0

# ok
```

`ember test` doesn't include a TAP version line now, but if it did in the future, all the lines above `ok 1` would be troublesome. If a consumer was permitted to ignore those lines before the TAP stream officially starts, I think the job of producers would be easier.

This might have ramifications to the strict pragma that is proposed.

Reply:
Ignoring unknowns, I think, is mainly for forward-compat and to reduce impact of unintended output. In terms of intentional diagonistic output like this, I'd expect e.g. Ember to not let these through. Either muting them, or forwarding them with `# ` ensured in front of it. If they are uncontrollable and intentionally let through, the coul also wreak havoc in more nominal ways if they happen to match TAP instructions.